### ISSUE TITLE
Request: Expose getImageSource()

### ISSUE THREAD

**Issue by andrioid** [expo/vector-icons#26]
Is it possible to expose getImageSource on the icon-set components?

I've tried adding it myself, but no luck yet.
Seems like the `NativeIconAPI` is not available, so I can't just add the method.

I would really appreciate a clarification and hopefully a fix.

I'll leave the doc link in the rnvi readme.
https://github.com/oblador/react-native-vector-icons#usage-as-png-imagesource-object

**Comment by brentvatne** [comment]
I can be convinced for why this is needed, but the only reason I can see is to use it with ToolBarAndroid or TabBarIOS, and I don't recommend using either of those, so we'll need more than that. Tagged as wontfix until then!

**Comment by andrioid** [comment]
Just for arguments sake, I'd like to hear why you wouldn't recommend using standard components from the react-native project. Some of us try to strive for that "native look" and that's just easier with actual native components..

My use-case is, as you mentioned TabBarIOS and ToolBarAndroid.

As for the 'wontfix' tag - that's fair. I imagine that the more native stuff you guys support, the more native code there is to maintain and most stuff _can_ be done in JS anyway.

Can I suggest a little foot-note to the icon-documentation mentioning that getImageSource is not supported and that you recommend against using TabBarIOS and ToolbarAndroid?

**Comment by brentvatne** [comment]
@andrioid - sure. I don't recommend using them because they are not cross-platform and can be easily re-created using react-native components, and they are less flexible even on the platform they do work on than just using a rebuilt one with react-native components. Is there some particular behavior you see in TabBarIOS and ToolBarAndroid that you don't get with just using the react-navigation tab bar and navigation bar? If so, we should improve react-navigation accordingly

**Comment by brentvatne** [comment]
Oops, forgot to mention the footnote -- yeah we should do this. I have had an open issue here for a while to warn or throw an error with TabBarIOS and vector-icons, if you could add a note to README and/or add some warning/error at runtime and send a PR this would be very helpful!

**Comment by dannin** [comment]
How do we use the icons in a MapView without getImageSource?

**Comment by brentvatne** [comment]
@dannin - use images i suppose

**Comment by oxyii** [comment]
+1, I need `getImageSource`

**Comment by klase** [comment]
Why is this tagged wontfix? How else would you use the icons on a map view?

**Comment by brentvatne** [comment]
use a static image @klase, i don't see why this needs to be done at runtime

**Comment by klase** [comment]
@brentvatne In our case we load the markers from a json file and since only local image resources are allowed to be used in react-native-maps, it seems odd to include assets twice when we already reference the icon set elsewhere (for tab navigation for example).

**Comment by brentvatne** [comment]
@klase - you can make it a remote image local by downloading it with `FileSystem.downloadAsync` :) so there is a workaround for now.

i definitely empathize with what you're saying from an ergonomics point of view though - it seems easier to just get the image source from the font at runtime. i'd be open to a proposal for a more generic module in expo to do this, but i don't think we should add it to the vector-icons lib and make it inherently coupled to icons.

**Comment by oxyii** [comment]
Looks like `@expo/vector-icons` does not include `react-native-vector-icons` **NATIVE** modules, so `@expo/vector-icons` cannot expose RNVI native methods... But `babel-preset-expo` [trying to tell us](https://github.com/expo/expo/blob/26f75ecfa4af95c4f835c3ecbc1f90eba043f651/packages/babel-preset-expo/index.js#L51-L53) that this is a complete analogue. Unfortunately, it is not. And I don't think that we will be able to convince expo members of the opposite

**Comment by brentvatne** [comment]
i think the best approach is to extract those native methods from react-native-vector-icons into another library that is specific to the functionality that they provide.

we don't like to include native modules when the api is designed as a grab bag of assorted functionality - we prefer if they are broken down into general purpose primitives. if we don't do this then the api surface area becomes a mess of slightly overlapping functionality in different libraries with a burden of maintaining ad-hoc implementations of the same thing in different places resting on us.

**Comment by brentvatne** [comment]
perhaps a good place to include this method could be on [expo-font](https://docs.expo.io/versions/latest/sdk/font/) - named something like `extractGlyph` - then we could expose `getImageSource` but have it depend on that implementation instead, and people can use the more generic `Font.extractGlyph` elsewhere too

**Comment by klase** [comment]
@brentvatne FileSystem.downloadAsync is not really a practical solution for us as it would require a network connection initially, but I do see your point about the api surface. +1 for adding it to expo-font.

**Comment by oxyii** [comment]
I still think it should be here. `react-native-vector-icons` have a **documented** behavior, `@expo/vector-icons` forced (via babel) `react-native-vector-icons` requests to `@expo/vector-icons`... According to my logic, native methods should be implemented here

**Comment by brentvatne** [comment]
@klase - if the image is already bundled in your app (via `assetBundlePatterns`) then you can use it just fine w/o `FileSystem.downloadAsync`, so that could be an option for you for now.

@oxyii - i updated the readme for you so that the documented behavior in @expo/vector-icons is clear https://github.com/expo/vector-icons/commit/233adbbdf7eb5b902371b7f3cff3d8784c59cc39

**Comment by oxyii** [comment]
@brentvatne ha-ha, thank you. But this is just escape from the problem. `babel-preset-expo` is still forced RNVI requests to `@expo/vector-icons` and `@expo/vector-icons` still does not contain RNVI documented functionality.

Look, I don’t really care anymore. Expo - is a **BEST**, really best way to learn react-native. And I am really very grateful to expo... Also I noticed that expo is moving towards modularity and it's great! But compatibility with native modules ceased to be enough for me... So I went to pure react-native.

But it seems that this question will still rise before you. Good luck to all!

**Comment by oxyii** [comment]
@iddan can you join this discussion? I think your [`react-native-canvas`](https://github.com/iddan/react-native-canvas) lib with merged [react-native-canvas#109](https://github.com/iddan/react-native-canvas/pull/109) can help workaround this

**Comment by iddan** [comment]
I'm sorry @oxyii but I'm not sure how can I help here. Would you mind giving me some context?

**Comment by oxyii** [comment]
@iddan `@expo/vector-icons` cannot provide [RNVI](https://github.com/oblador/react-native-vector-icons) method `getImageSource` because this method is native. But this method just create 2d context, print one symbol (with specific font) into it and then get base64 string with png MIME...
We need to this method because native components (such a maps of bottom tabbar) cannot use vector

**Comment by iddan** [comment]
I'm not sure creating a canvas for this would be the right approach. But I was intending to use Skia's Canvas Kit in React Native Canvas instead of a WebView. Maybe it can help you achieve what you want: https://skia.org/user/modules/canvaskit

**Comment by oxyii** [comment]
@iddan Very interesting... but looks like WebAssembly is not yet supported on RN... at least not on all platforms and simulators https://github.com/react-native-community/jsc-android-buildscripts/issues/113

Seems that this issue can not be resolved without a WebView. Sorry guys.

> [`react-native-canvas`](https://github.com/iddan/react-native-canvas) lib with merged [react-native-canvas#109](https://github.com/iddan/react-native-canvas/pull/109) can help workaround this...

Looks like this is the only solution for now

**Comment by oxyii** [comment]
Despite the fact that this functionality is not relevant for me, I could not help but think about it.
@andrioid @klase @dannin @robksawyer @foreleven the only thing I could do for you was server conversion https://github.com/oxyii/getimagesource

if the expo does not plan to include RNVI native module then I think this issue can be closed

**Comment by iddan** [comment]
@oxyii you can compile WASM to asm.js to support JSC

**Comment by oxyii** [comment]
@iddan I'm not an expert on emsdk compilation tuning... Pre-built [wasm](https://www.npmjs.com/package/canvaskit-wasm) have a size about 7Mb. My asm version - more that 13Mb... too expensive for this issue in both cases.. even if it will work )))

**Comment by nandorojo** [comment]
I want the image source so that I can render it in a native iOS menu component. There's no way to render JSX inside of this component. Is the solution to just copy the src file into my `assets`?

**Comment by oxyii** [comment]
@nandorojo  a lot of time has passed since then. I'm sure you can use [pre-caching](https://docs.expo.dev/guides/preloading-and-caching-assets/#pre-loading-and-caching-assets). But I do not recommend using my deployments of getImageSource repo. I do not guarantee that the resources will work. Feel free to clone repo and create your own deployment. But it's easier to copy the image files into `assets`

**Comment by brentvatne** [comment]
using eas build you can use the getImageSource function from react-native-vector-icons

**Comment by mpinkard** [comment]
> using eas build you can use the getImageSource function from react-native-vector-icons

@brentvatne Would you mind elaborating on your comment?

**Comment by brentvatne** [comment]
you can learn more about eas build vs expo build in these blog posts:

https://blog.expo.dev/expo-managed-workflow-in-2021-5b887bbf7dbb
https://blog.expo.dev/expo-managed-workflow-in-2021-d1c9b68aa10

tl;dr expo build doesn't let you include custom native code, eas build does. you can also get a similar experience to expo go by [using expo-dev-client](https://docs.expo.dev/development/introduction/)

**Comment by nandorojo** [comment]
probably best to just download and import the image of the icon itself

**Comment by artsnr1** [comment]
How to do that? except using Filesystem.download...?

**Comment by nandorojo** [comment]
just download the icon from figma or your icon maker’s website

**Comment by kevindetry** [comment]
Looks like `react-native-bottom-tabs` needs this: https://okwasniewski.github.io/react-native-bottom-tabs/docs/guides/usage-with-vector-icons.html
